source/request-response/request_response_client.c

```diff
@@ -0,0 +1,388 @@
+
+    struct aws_rr_subscription_manager_options subscription_manager_options = {
+        .operation_timeout_seconds = rr_client->config.operation_timeout_seconds,
+        .max_subscriptions = rr_client->config.max_subscriptions,
+        .subscription_status_callback = s_aws_rr_client_subscription_status_event_callback,
+        .userdata = rr_client,
+    };
+
+    aws_rr_subscription_manager_init(
+        &rr_client->subscription_manager, allocator, rr_client->client_adapter, &subscription_manager_options);
+}
+
+static void s_mqtt_request_response_client_initialize_task_fn(
+    struct aws_task *task,
+    void *arg,
+    enum aws_task_status task_status) {
+    (void)task;
+
+    AWS_FATAL_ASSERT(task_status != AWS_TASK_STATUS_CANCELED);
+
+    struct aws_mqtt_request_response_client *client = arg;
+
+    if (client->state == AWS_RRCS_UNINITIALIZED) {
+        s_aws_rr_client_init_subscription_manager(client, client->allocator);
+
+        client->state = AWS_RRCS_ACTIVE;
+    }
+
+    if (client->config.initialized_callback != NULL) {
+        (*client->config.initialized_callback)(client->config.user_data);
+    }
+
+    /* give up the internal ref we held while the task was pending */
+    aws_ref_count_release(&client->internal_ref_count);
+}
+
+static void s_setup_cross_thread_initialization(struct aws_mqtt_request_response_client *rr_client) {
+    /* now that it exists, 1 internal ref belongs to protocol adapter termination */
+    aws_ref_count_acquire(&rr_client->internal_ref_count);
+
+    /* 1 internal ref belongs to the initialize task until it runs */
+    aws_ref_count_acquire(&rr_client->internal_ref_count);
+    aws_task_init(
+        &rr_client->initialize_task,
+        s_mqtt_request_response_client_initialize_task_fn,
+        rr_client,
+        "mqtt_rr_client_initialize");
+    aws_event_loop_schedule_task_now(rr_client->loop, &rr_client->initialize_task);
+}
+
+struct aws_mqtt_request_response_client *aws_mqtt_request_response_client_new_from_mqtt311_client(
+    struct aws_allocator *allocator,
+    struct aws_mqtt_client_connection *client,
+    const struct aws_mqtt_request_response_client_options *options) {
+
+    struct aws_mqtt_request_response_client *rr_client =
+        s_aws_mqtt_request_response_client_new(allocator, options, aws_mqtt_client_connection_get_event_loop(client));
+
+    struct aws_mqtt_protocol_adapter_options adapter_options = {
+        .subscription_event_callback = s_aws_rr_client_protocol_adapter_subscription_event_callback,
+        .incoming_publish_callback = s_aws_rr_client_protocol_adapter_incoming_publish_callback,
+        .terminate_callback = s_aws_rr_client_protocol_adapter_terminate_callback,
+        .connection_event_callback = s_aws_rr_client_protocol_adapter_connection_event_callback,
+        .user_data = rr_client,
+    };
+
+    rr_client->client_adapter = aws_mqtt_protocol_adapter_new_from_311(rr_client->allocator, &adapter_options, client);
```

Comment:
`s_aws_mqtt_request_response_client_new` can return null if `sm_options` fails validation. We need a null check on the `rr_client` before assigning a `client_adapter`